AssemblyProvider.cs

```diff
@@ -0,0 +1,274 @@
+            throw new ArgumentException($"{Path.GetFileName(modulePath)} does not end in .exe");
+
+        var pdbPath = Path.Combine(Path.GetDirectoryName(modulePath)!, Path.GetFileNameWithoutExtension(modulePath) + ".pdb");
+        if (!File.Exists(pdbPath))
+            throw new ArgumentException($"{Path.GetFileName(modulePath)} does not have a companion .pdb file.");
+
+        pe = PEFile.FromFile(modulePath);
+        pdb = new PdbFileReader(pdbPath);
+
+        text = pe.Sections.Single(e => e.Name.Equals(".text"));
+    }
+
+    public OSPlatform Platform => OSPlatform.Windows;
+
+    public IEnumerable<Architecture> Architectures() {
+        yield return pe.FileHeader.Machine switch {
+            MachineType.Amd64 => Architecture.X64,
+            MachineType.Arm64 => Architecture.Arm64,
+            _ => throw new ArgumentException($"Unknown PE architecture type. {Enum.GetName(typeof(MachineType), pe.FileHeader.Machine)}")
+        };
+    }
+
+    public Range FunctionFileRange(Architecture arch, string fName) {
+        Debug.Assert(Architectures().Contains(arch), "Tried to patch for unknown architecture.");
+
+        var func = pdb.Functions.Single(e => e.Name.Equals(fName));
+        var fnOffset = (int)pe.RvaToFileOffset(func.Offset + text.Rva);
+        return new Range(fnOffset, fnOffset + (int)func.CodeSize);
+    }
+
+    public void Dispose() {
+        pdb.Dispose();
+    }
+
+}
+
+public sealed class LinuxAssemblyProvider : AssemblyProvider {
+
+    private readonly ELF<ulong>         elf;
+    private readonly int                symOffset;
+    private readonly SymbolTable<ulong> symTable;
+
+    public LinuxAssemblyProvider(ELF<ulong> elf) {
+        this.elf = elf;
+
+        var text = (ProgBitsSection<ulong>)elf.GetSection(".text");
+        symOffset = (int)(text.Offset - text.LoadAddress);
+        symTable = (SymbolTable<ulong>)elf.GetSection(".symtab");
+    }
+
+    public OSPlatform Platform => OSPlatform.Linux;
+
+    public IEnumerable<Architecture> Architectures() {
+        yield return elf.Machine switch {
+            ElFMachine.AMD64 => Architecture.X64,
+            ElFMachine.AArch64 => Architecture.Arm64,
+            _ => throw new ArgumentException($"Unknown ELF architecture type. {Enum.GetName(typeof(ElFMachine), elf.Machine)}")
+        };
+    }
+
+    public Range FunctionFileRange(Architecture arch, string fName) {
+        Debug.Assert(Architectures().Contains(arch), "Tried to patch for unknown architecture.");
+
+        var func = symTable.Entries.Single(e => e.Name.Equals(fName));
+        var fnOffset = (int)func.Value + symOffset;
+        return new Range(fnOffset, fnOffset + (int)func.Size);
+    }
+
+    public void Dispose() {
+        elf.Dispose();
+    }
+
+}
+
+public sealed class MacosAssemblyProvider : AssemblyProvider {
+
+    private Dictionary<Architecture, MachO64> binaries;
+    private Signer?                           _signer;
```

Comment:
haha, I won't make you change it, but this spacing is very non-standard

Reply:
Rider default formatting rules aligns things. Have disabled it and formatted the file.